### PR TITLE
Attendance UI

### PR DESCRIPTION
**Pull request by bethbrains**
# Critical Changes

# Changes
- Add a UI for gathering attendance data on a Service Session page. (No save function yet.)

# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
- [x] Place holder data is incorporated (Refer to [Case Management Sample Data document](https://quip.com/cbFcAPJF8t0z))
- [x] Base axe-core JEST test included for any new LWC (No critical violations)
- [x] CRUD/FLS is enforced in Apex
- [x] Permission sets & field sets are updated to account for CRUD/FLS/TABS for new object metadata
- [x] All modified classes are unit tested (Apex) at 100% coverage
- [x] UX approval or UX not necessary
- [x] PR contains draft release notes
- [x] Attach work item to the pull request with [Lurch](https://salesforce.quip.com/50ZRA5LEWVzH) `**Lurch:attach W-000000` or `**Lurch:add`
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [ ] QE story level testing completed

### CODE REVIEW

**Comment by bethbrains** [comment]
[W-8125650](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000089herIAA/view)

**Comment by bethbrains** [comment]
Note to QE: @kselvocki @rmonica08, this does not yet fully support field sets, even though the corresponding field set exists. there is a separate story for testing it with field set-driven fields.

**Comment by bethbrains** [comment]
[W-8126133](https://gus.lightning.force.com/a07B00000089hp1IAA)

File: force-app/main/default/lwc/attendance/attendance.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                field.size = 1;
+            } else if (LONG_DATA_TYPES.includes(field.type)) {
+                field.size = 3;
+            } else {
+                field.size = 2;
+            }
+
+            if (field.apiName === this.fields.contact.fieldApiName) {
+                field.isContactField = true;
+                field.size = 2;
```
Comment (bmuniswamy) [inline force-app/main/default/lwc/attendance/attendance.js:118]: @bethbrains , Isn't this the lookup field which should have a size of 3 and is set on line 111. Should we be setting the size again here.

File: force-app/main/default/lwc/attendanceRow/attendanceRow.html [hunk 2]
```diff
@@ -0,0 +1,80 @@
+<template>
+    <div
+        class="slds-box small-box slds-var-m-around_x-small slds-var-p-horizontal_medium"
+    >
+        <lightning-record-edit-form
+            record-id={recordId}
+            object-api-name={serviceDeliveryObject}
+        >
+            <!-- onload={handleLoad}
```
Comment (bmuniswamy) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.html:9]: Do we need to keep the commented lines.

File: force-app/main/default/lwc/attendance/attendance.html [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                            field-set={fieldSet}
+                            unit-of-measurement={unitOfMeasurement}
+                            key={delivery.index}
+                        ></c-attendance-row>
+                    </template>
+                </template>
+            </div>
+        </div>
+
+        <div slot="footer" class="slds-text-align_right">
```
Comment (bmuniswamy) [inline force-app/main/default/lwc/attendance/attendance.html:25]: So when there are no participant records and we look at the attendence component in the UI, we see a blank screen with the Submit button. I am assuming that we should not be seeing the submit button. Would that be addressed in the following stories

**Review by bmuniswamy — COMMENTED** [review]
Just a couple of follow up questions

File: .gitignore [hunk 4]
```diff
@@ -57,6 +57,7 @@ yarn-error.log*
 # LWC
 **/lwc/jsconfig.json
 **/lwc/.eslintrc.json
+coverage/
```
Comment (bmuniswamy) [inline .gitignore:60]: Beth, why is this needed. Do we want to ignore the code coverage.

File: force-app/main/default/lwc/attendanceRow/attendanceRow.css [hunk 5]
```diff
@@ -0,0 +1,3 @@
+.small-box {
+    padding: var(--lwc-varSpacingXxxSmall, 0.125rem);
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.css:2]: Nice!

File: force-app/main/default/lwc/attendanceRow/attendanceRow.html [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                onsuccess={handleSuccess}
+                onsubmit={handleSubmit}
+                onerror={handleSaveError} -->
+            <lightning-layout multiple-rows="true" horizontal-align="left">
+                <template for:each={fieldSet} for:item="field">
+                    <lightning-layout-item
+                        size="12"
+                        small-device-size="6"
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.html:17]: I wonder if we want to try and anchor the fields to the bottom or truncate the label or something, when you shrink the screen the wrapping of text on one field makes the alignment off on the others. Let me know what you think... 
Screen Shot 2020-11-06 at 8 26 01 AM

File: force-app/main/default/lwc/attendanceRow/attendanceRow.html [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                                    onchange={handleQuantityChange}
+                                    type="number"
+                                    min="0"
+                                    value={quantity}
+                                ></lightning-input>
+                            </div>
+                        </template>
+                        <template if:true={field.isNormalInputField}>
+                            <lightning-input-field
+                                id={field.apiName}
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.html:49]: Can we use id's in lwc?

File: force-app/main/default/lwc/attendanceRow/attendanceRow.html [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                                field-name={field.apiName}
+                                onchange={handleInputChange}
+                                disabled={field.disabled}
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.html:52]: Are we ever setting this to true?

File: force-app/main/default/lwc/attendanceRow/attendanceRow.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
+import { LightningElement, api, track, wire } from "lwc";
+import { getRecord } from "lightning/uiRecordApi";
+import SERVICEDELIVERY_OBJECT from "@salesforce/schema/ServiceDelivery__c";
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.js:3]: Missing an _ between service and delivery

File: force-app/main/default/lwc/attendanceRow/attendanceRow.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    get recordId() {
+        return this.record.Id;
+    }
+
+    get quantity() {
+        return this.record[this.fields.quantity.fieldApiName];
+    }
+
+    @wire(getRecord, {
+        recordId: "$contactId",
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.js:58]: I am still wondering if we are better off getting everything in Apex in one trip and passing it down. Maybe a good topic for dev sync?

File: force-app/main/default/classes/AttendanceSelector.cls [hunk 11]
```diff
@@ -29,7 +32,12 @@ public with sharing class AttendanceSelector {
         }
 
         List<ServiceParticipant__c> serviceParticipants = [
-            SELECT Id, Contact__c, Service__c, ProgramEngagement__c
+            SELECT
```
Comment (bmuniswamy) [inline force-app/main/default/classes/AttendanceSelector.cls:35]: @bethbrains ,I think we need to add an order by clause since one of the AC in the story says that the contact names should be listed alphabetically

File: force-app/main/default/lwc/attendanceRow/attendanceRow.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        fields: [CONTACT_NAME_FIELD],
+    })
+    wiredDelivery(result) {
+        if (!(result.data || result.error)) {
+            return;
+        }
+        if (result.data) {
+            this.name = result.data.fields.Name.value;
+            console.log("here");
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.js:67]: Marco :)

**Review by jjbennett — COMMENTED** [review]
Passing along the comments, while I take this for a test drive. Great work, looking really good!

**Review by bmuniswamy — COMMENTED** [review]
I have a couple of comments. One more thing I noticed in the UI is that when I am in the last row of the attendence UI and when I click on the picklist I am unable to see the picklist values until I scroll down. I think that needs to be fixed.

File: force-app/main/default/lwc/attendance/attendance.js [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            return;
+        }
+
+        if (result.data) {
+            let schedule = getChildObjectByName(result.data.fields, "ServiceSchedule__r");
+            let service = getChildObjectByName(schedule.value.fields, "Service__r");
+            this.unitOfMeasurement =
+                service.value.fields[UNITMEASUREMENT_SERVICE_FIELD.fieldApiName].value;
+        } else if (result.error) {
+            handleError(result.error);
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendance/attendance.js:58]: Can we console.log this here, otherwise when an error occurs they will get a toast even if they haven't interacted with our component. On that note if they remove the contact from the field set an error occurs, do we want to manually require that field. Also we could run into that situation if the user doesn't have access to that field.

File: force-app/main/default/classes/AttendanceSelector.cls (hunk 11 above)
Comment (jjbennett) [reply force-app/main/default/classes/AttendanceSelector.cls:35]: Nice catch, but I would suggest we do our sorting in JS vs SOQL

**Comment by jjbennett** [comment]
[W-8125664](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000089hfVIAQ/view)

File: force-app/main/default/lwc/attendanceRow/attendanceRow.html (hunk 2 above)
Comment (AndersonTarren) [reply force-app/main/default/lwc/attendanceRow/attendanceRow.html:18]: Agree, I don't see a reason to keep commented lines here.

File: force-app/main/default/lwc/attendanceRow/attendanceRow.html (hunk 6 above)
Comment (AndersonTarren) [reply force-app/main/default/lwc/attendanceRow/attendanceRow.html:26]: Thumbs up for this idea - would be cool to find a way to keep things looking nice, even when compact.

File: force-app/main/default/lwc/attendanceRow/attendanceRow.js [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+export default class AttendanceRow extends LightningElement {
+    @track localRecord;
+    @track localFieldSet;
+
+    @api unitOfMeasurement;
+    @api presentStatus = PRESENT_STATUS;
+    @api readOnly = false;
+    name;
+    _isEdited;
+    recordId;
```
Comment (AndersonTarren) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.js:31]: Probably not a big deal, but caught my eye that we're using recordId here, but not in the typical @api context.

File: force-app/main/default/lwc/participantSelector/participantSelector.css [hunk 15]
```diff
@@ -1,3 +1,12 @@
+/*
```
Comment (AndersonTarren) [inline force-app/main/default/lwc/participantSelector/participantSelector.css:1]: Duplicate licensing being applied here.

**Review by AndersonTarren — CHANGES_REQUESTED** [review]
Couple items here to adjust - the duplicate heading and the commented code Bharathi called out.

**Review by lbeller — COMMENTED** [review]
A couple of minor changes:

-Successfully saved {0} attendance records.  = Saved {0} attendance records.

-There are Service Delivery rows with errors. = Not sure where this is in the UI?

- There were no edited attendance records to update. = There are no attendance records to update. 

- You don't have access to insert or update this record. Your Salesforce admin can help with that. = You can't add or update this record. Your Salesforce admin can help with that.

File: force-app/main/default/lwc/participantSelector/participantSelector.css (hunk 15 above)
Comment (bethbrains) [reply force-app/main/default/lwc/participantSelector/participantSelector.css]: cool, something wrong with my copyright script!
